### ISSUE TITLE
'501 Not Implemented' is an improper response for available API versions

### ISSUE THREAD

**Issue by ocket8888** [apache/trafficcontrol#2997]
HTTP spec designates the 501 response code to be used for a lack of _HTTP_ functionality. That more or less means that the HTTP method used in the request isn't implemented by the server (*not* that the method is not available for an existing resource - that would entail a 405 response). The proper response code for a path that does not exist on the server is `404 Not Found`.

Maybe we could use the `alerts` JSON object instead?

Alternatively, this would also be resolved if #2872 were resolved.

**Comment by rob05c** [comment]
@ocket8888 The RFC you linked, RFC 2616, has been obsoleted by RFC 7231. Per RFC 7231: 

>   The 501 (Not Implemented) status code indicates that the server does
>   not support the functionality required to fulfill the request.  This
>   is the appropriate response when the server does not recognize the
>   request method and is not capable of supporting it for any resource.

https://tools.ietf.org/html/rfc7231#section-6.6.2

Yes, 501 is appropriate for methods not supported. But it also more generically "indicates that the server does not support the functionality required to fulfill the request". Nowhere in there does the RFC indicate the functionality not supported SHOULD or MUST be "HTTP functionality".

>The proper response code for a path that does not exist on the server is 404 Not Found.

A 404 is technically appropriate, but I do not believe a 501 to be inappropriate. A 404 does not indicate that the functionality may or may not be supported by future versions, where 501 allows us to indicate the real problem, namely that newer API version was requested. Yes, we could encode the real problem in a proprietary message in the body. but a 501 is the most likely to more readily help the user understand the real problem.

https://tools.ietf.org/html/rfc7231#section-6.6
>   The 5xx (Server Error) class of status code indicates that the server
>   is aware that it has erred or is incapable of performing the
>   requested method.

https://tools.ietf.org/html/rfc7231#section-6.5
>    The 4xx (Client Error) class of status code indicates that the client
>   seems to have erred.

A 4xx indicates a client problem, where a 5xx indicates a server problem. In this case, the client is correct, it's just newer than the server. The server is outdated. The real problem is that to use the new client functionality, the user needs to upgrade their Traffic Ops server. So, a 5xx, and specifically a 501, most accurately conveys the real problem to the user. 

Yes, you could argue the client has "erred" by being too new. But that seems like an odd argument to me.

TLDR Yes, we're using a broad interpretation of the RFC, but I don't believe it to be strictly incorrect, and it's the best way to quickly convey the problem to the user.

**Comment by ocket8888** [comment]
I disagree that the problem is server-side, the client requested access to a resource that does not exist - presumably the request failed because of that, not because the server itself failed to accomplish anything.

> The 404 (Not Found) status code indicates that the origin server did not find a current representation for the target resource or is not willing to disclose that one exists.

**Comment by rawlinp** [comment]
Since we already use `404 Not Found` to indicate that the requested _resource_ does not exist (e.g. no server with id 4 exists), it probably makes sense to use `501` just to differentiate those kinds of errors from "my client is too new and is requesting an API version that the server doesn't support yet" errors.
